contracts/external/cw-fund-distributor/src/contract.rs

```diff
@@ -0,0 +1,628 @@
+        .add_attribute("voting_contract", voting_contract)
+        .add_attribute("total_power", total_power.power))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(cw20::Cw20ReceiveMsg {
+            sender: _,
+            amount,
+            msg: _,
+        }) => execute_fund_cw20(deps, env, info.sender, amount),
+        ExecuteMsg::FundNative {} => execute_fund_native(deps, env, info),
+        ExecuteMsg::ClaimCW20 { tokens } => execute_claim_cw20s(deps, env, info.sender, tokens),
+        ExecuteMsg::ClaimNatives { denoms } => {
+            execute_claim_natives(deps, env, info.sender, denoms)
+        }
+        ExecuteMsg::ClaimAll {} => execute_claim_all(deps, env, info.sender),
+    }
+}
+
+pub fn execute_fund_cw20(
+    deps: DepsMut,
+    env: Env,
+    token: Addr,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    let funding_deadline = FUNDING_PERIOD_EXPIRATION.load(deps.storage)?;
+    // if current block indicates claiming period, return an error
+    if funding_deadline.is_expired(&env.block) {
+        return Err(ContractError::FundDuringClaimingPeriod {});
+    }
+
+    let balance = CW20_BALANCES.may_load(deps.storage, token.clone())?;
+
+    match balance {
+        Some(old_amount) => CW20_BALANCES.save(
+            deps.storage,
+            token.clone(),
+            &old_amount.checked_add(amount).unwrap(),
+        )?,
+        None => CW20_BALANCES.save(deps.storage, token.clone(), &amount)?,
+    }
+
+    Ok(Response::default()
+        .add_attribute("method", "fund_cw20")
+        .add_attribute("token", token)
+        .add_attribute("amount", amount))
+}
+
+pub fn execute_fund_native(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    let funding_deadline = FUNDING_PERIOD_EXPIRATION.load(deps.storage)?;
+    // if current block indicates claiming period, return an error
+    if funding_deadline.is_expired(&env.block) {
+        return Err(ContractError::FundDuringClaimingPeriod {});
+    }
+
+    // collect a list of successful funding kv pairs
+    let attributes: Vec<(String, String)> = info
+        .funds
+        .into_iter()
+        .filter(|coin| coin.amount > Uint128::zero())
+        .map(|coin| {
+            let balance = NATIVE_BALANCES
+                .may_load(deps.storage, coin.denom.clone())
+                .unwrap_or_default();
+            match balance {
+                Some(old_amount) => NATIVE_BALANCES.save(
+                    deps.storage,
+                    coin.denom.clone(),
+                    &old_amount.checked_add(coin.amount).unwrap(),
+                ),
+                None => NATIVE_BALANCES.save(deps.storage, coin.denom.clone(), &coin.amount),
+            }
+            .unwrap();
```

Comment:
Would be nice to do this without `unwrap()`s.

Reply:
```suggestion
            match balance {
                Some(old_amount) => NATIVE_BALANCES.save(
                    deps.storage,
                    coin.denom.clone(),
                    &old_amount.checked_add(coin.amount).unwrap(),
                ),
                None => NATIVE_BALANCES.save(deps.storage, coin.denom.clone(), &coin.amount),
            }?;
```

Reply:
> Would be nice to do this without `unwrap()`s.

I tried but the issue is with `?` operator only being usable with `Result` or `Option`. I feel like the `checked_add` should never overflow given the `U128` range, but I suppose we could do something like this instead?
```rust
    let mut attributes: Vec<(String, String)> = Vec::new();
    for coin in info.funds {
        if coin.amount > Uint128::zero() {
            let balance = NATIVE_BALANCES
                .may_load(deps.storage, coin.denom.clone())
                .unwrap_or_default();

            let mut new_amount = coin.amount;
            // add any previous balances
            if let Some(previous_amount) = balance {
                new_amount.checked_add(previous_amount)
                    .map_err(|e| ContractError::Std(StdError::from(e)))?;
            };
            NATIVE_BALANCES.save(
                deps.storage,
                coin.denom.clone(),
                &new_amount,
            )?;
        }
    }

```

Reply:
sometimes imperative > functional. this looks great